### PR TITLE
Changes for PR2 of BlameModeForDump

### PR DESCRIPTION
**Pull request by karanjitsingh**
Goal of this PR
* To add ProcessDumpUtility to BlameDataCollector

Extras
* Rename TestHostInitialized to TestHostLaunched
* Deleted ProcessHelperTests

The final PR3 will contain full integration and enable dump mode with changes to EnableBlameArgumentProcessor

### CODE REVIEW

**Comment by karanjitsingh** [comment]
@dotnet-bot test Windows_NT / Release Build

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            {
+                return dumpFile;
+            }
+
+            if (EqtTrace.IsInfoEnabled)
+            {
+                EqtTrace.Verbose(string.Format(CultureInfo.InvariantCulture, "ProcessDumpUtility: GetCrashDumpFile: No dump file generated."));
+            }
+
+            return string.Empty;
```
Comment (smadala) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs]: Throw exception if fail to create dump.

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs [hunk 2]
```diff
@@ -133,15 +151,59 @@ private void SessionEnded_Handler(object sender, SessionEndEventArgs args)
...
+                if (!dumpFile.Equals(string.Empty))
+                {
+                    var fileTranferInformation = new FileTransferInformation(this.context.SessionDataCollectionContext, dumpFile, true);
+                    this.dataCollectionSink.SendFileAsync(fileTranferInformation);
+                }
+                else
+                {
+                    if (EqtTrace.IsWarningEnabled)
+                    {
+                        EqtTrace.Warning("blame:CollectDump was enabled but dump file was not generated.");
```
Comment (smadala) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs]: Nit: Add classname and function to log.

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs [hunk 3]
```diff
@@ -133,15 +151,59 @@ private void SessionEnded_Handler(object sender, SessionEndEventArgs args)
...
+
+            try
+            {
+                this.processDumpUtility.StartProcessDump(args.TestHostProcessId, this.attachmentGuid, this.GetResultsDirectory());
+            }
+            catch (Exception e)
+            {
+                if (EqtTrace.IsWarningEnabled)
+                {
+                    EqtTrace.Warning(string.Format(CultureInfo.InvariantCulture, "BlameCollector: TestHostLaunched_Handler: Could not start process dump. {0}", e.Message));
```
Comment (smadala) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs]: Print callstack.

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameLogger.cs [hunk 4]
```diff
@@ -83,6 +83,21 @@ public void Initialize(TestLoggerEvents events, string testRunDictionary)
             events.TestRunComplete += this.TestRunCompleteHandler;
         }
 
+        /// <summary>
+        /// Initializes the Logger.
+        /// </summary>
+        /// <param name="events">Events that can be registered for.</param>
+        /// <param name="testRunDictionary">Test Run Directory</param>
+        public void Initialize(TestLoggerEvents events, Dictionary<string, string> testRunDictionary)
```
Comment (smadala) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameLogger.cs]: Looks like this is not required.

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+            this.processHelper.WaitForProcessExit(this.procDumpProcess);
+
+            var dumpFile = Path.Combine(this.testResultsDirectory, this.dumpFileName);
+            if (this.fileHelper.Exists(dumpFile))
+            {
+                return dumpFile;
+            }
+
+            if (EqtTrace.IsInfoEnabled)
```
Comment (smadala) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs]: Nit; Make this error.

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs [hunk 6]
```diff
@@ -252,14 +252,14 @@ public Collection<AttachmentSet> SessionEnded(bool isCancelled = false)
         }
 
         /// <inheritdoc/>
-        public void TestHostInitialized(TestHostInitializedEventArgs testHostInitializedEventArgs)
+        public void TestHostLaunched(TestHostLaunchedEventArgs TestHostLaunchedEventArgs)
```
[the remark this answers is not in the document]
Comment (karanjitsingh) [reply src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionManager.cs]: done

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs [hunk 7]
```diff
@@ -42,9 +47,13 @@ public BlameCollector()
         /// <param name="blameReaderWriter">
         /// BlameReaderWriter instance.
         /// </param>
-        protected BlameCollector(IBlameReaderWriter blameReaderWriter)
+        /// <param name="processDumpUtility">
+        /// ProcessDumpUtility instance.
+        /// </param>
+        protected BlameCollector(IBlameReaderWriter blameReaderWriter, IProcessDumpUtility processDumpUtility)
```
[the remark this answers is not in the document]
Comment (karanjitsingh) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs]: done

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs (hunk 2 above)
Comment (karanjitsingh) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs]: done

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs (hunk 3 above)
Comment (karanjitsingh) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs]: done

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameLogger.cs (hunk 4 above)
Comment (karanjitsingh) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameLogger.cs]: removed

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs (hunk 5 above)
Comment (karanjitsingh) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs]: done

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+        /// <summary>
+        /// Get procdump executable path
+        /// </summary>
+        /// <returns>procdump executable path</returns>
+        private string GetProcDumpExecutable()
+        {
+            var procdumpPath = Environment.GetEnvironmentVariable("PROCDUMP_PATH");
+            if (!string.IsNullOrEmpty(procdumpPath))
+            {
```
[the remark this answers is not in the document]
Comment (karanjitsingh) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs]: added

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+                if (this.environment.Architecture == PlatformArchitecture.X64)
+                {
+                    filename = "procdump64.exe";
+                }
+                else if (this.environment.Architecture == PlatformArchitecture.X86)
+                {
+                    filename = "procdump.exe";
+                }
+
+                var procDumpExe = Path.Combine(procdumpPath, "Extensions", filename);
```
[the remark this answers is not in the document]
Comment (karanjitsingh) [reply src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs]: removed

**Review by smadala — APPROVED** [review]

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs [hunk 10]
```diff
@@ -133,15 +151,59 @@ private void SessionEnded_Handler(object sender, SessionEndEventArgs args)
...
+                var filepath = Path.Combine(this.GetResultsDirectory(), Constants.AttachmentFileName + "_" + this.attachmentGuid);
                 filepath = this.blameReaderWriter.WriteTestSequence(this.testSequence, filepath);
                 var fileTranferInformation = new FileTransferInformation(this.context.SessionDataCollectionContext, filepath, true);
                 this.dataCollectionSink.SendFileAsync(fileTranferInformation);
             }
 
+            if (this.processDumpEnabled)
+            {
+                var dumpFile = this.processDumpUtility.GetDumpFile();
+                if (!dumpFile.Equals(string.Empty))
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs]: Nit: string.IsNullorEmpty.

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameLogger.cs [hunk 11]
```diff
@@ -143,7 +143,9 @@ private IEnumerable<string> GetFaultyTestCaseNames(TestRunCompleteEventArgs e)
             {
                 if (attachmentSet.DisplayName.Equals(Constants.BlameDataCollectorName))
                 {
-                    var uriDataAttachment = attachmentSet.Attachments.LastOrDefault();
+                    // Process only Sequence_<GUID>.xml attachments
+                    var uriDataAttachment = attachmentSet.Attachments.LastOrDefault((attachment) => attachment.Uri.ToString().EndsWith(".xml"));
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameLogger.cs]: Add test

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+        /// <param name="processId">
+        /// Process Id
+        /// </param>
+        /// <param name="filename">
+        /// Filename for dump file
+        /// </param>
+        /// <returns>Arguments</returns>
+        private static string BuildProcDumpArgs(int processId, string filename)
+        {
+            return "-t -g -ma " + processId + " " + filename;
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs]: comments please

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+        }
+
+        /// <summary>
+        /// Get procdump executable path
+        /// </summary>
+        /// <returns>procdump executable path</returns>
+        private string GetProcDumpExecutable()
+        {
+            var procdumpPath = Environment.GetEnvironmentVariable("PROCDUMP_PATH");
+            if (!string.IsNullOrEmpty(procdumpPath))
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs]: >IsNullOrEmpty [](start = 24, length = 13)

Nit: whitespace

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/Interfaces/IProcessDumpUtility.cs [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+        string GetDumpFile();
+
+        /// <summary>
+        /// Launch procdump process
+        /// </summary>
+        /// <param name="processId">
+        /// Process ID of test host
+        /// </param>
+        /// <param name="dumpFileGuid">
+        /// Posfix for dump file, testhost.exe_<postfix>.dmp</guid>
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/Interfaces/IProcessDumpUtility.cs]: >Posfix [](start = 12, length = 6)

nit

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/Resources/Resources.Designer.cs [hunk 15]
```diff
@@ -68,5 +68,17 @@ internal class Resources {
                 return ResourceManager.GetString("AbortedTestRun", resourceCulture);
             }
         }
+
+
+        /// <summary>
+        ///   Looks up a localized string ProcDumpEnvVarEmpty
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/Resources/Resources.Designer.cs]: nit

File: src/Microsoft.TestPlatform.PlatformAbstractions/common/System/ProcessHelper.cs [hunk 16]
```diff
@@ -192,5 +192,15 @@ public string GetNativeDllDirectory()
 
             return Path.Combine(this.GetCurrentProcessLocation(), this.GetCurrentProcessArchitecture().ToString().ToLower());
         }
+
+        /// <inheritdoc/>
+        public void WaitForProcessExit(object process)
+        {
+            var proc = process as Process;
+            if (!proc.HasExited)
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.PlatformAbstractions/common/System/ProcessHelper.cs]: null check

File: test/Microsoft.TestPlatform.Extensions.BlameDataCollector.UnitTests/ProcessDumpUtilityTests.cs [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+        /// Initializes a new instance of the <see cref="ProcessDumpUtilityTests"/> class.
+        /// </summary>
+        public ProcessDumpUtilityTests()
+        {
+            this.mockFileHelper = new Mock<IFileHelper>();
+            this.mockProcessHelper = new Mock<IProcessHelper>();
+            this.mockProcDumpProcess = new Mock<Process>();
+            this.mockPlatformEnvironment = new Mock<IEnvironment>();
+
+            Environment.SetEnvironmentVariable("PROCDUMP_PATH", "procdump_folder");
```
Comment (singhsarab) [inline test/Microsoft.TestPlatform.Extensions.BlameDataCollector.UnitTests/ProcessDumpUtilityTests.cs]: Cleanup

File: test/Microsoft.TestPlatform.Extensions.BlameDataCollector.UnitTests/ProcessDumpUtilityTests.cs [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+        }
+
+        /// <summary>
+        /// GetDumpFile will return empty list of strings if no dump files found
+        /// </summary>
+        [TestMethod]
+        public void GetDumpFileWillReturnEmptyIfNoDumpFile()
```
Comment (singhsarab) [inline test/Microsoft.TestPlatform.Extensions.BlameDataCollector.UnitTests/ProcessDumpUtilityTests.cs]: >GetDumpFileWillReturnEmptyIfNoDumpFile [](start = 20, length = 38)

When will this happen, we should log this case?

File: test/Microsoft.TestPlatform.Extensions.BlameDataCollector.UnitTests/ProcessDumpUtilityTests.cs [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+            processDumpUtility.StartProcessDump(12345, "guid", "D:\\TestResults");
+
+            Assert.AreEqual(string.Empty, processDumpUtility.GetDumpFile());
+        }
+
+        /// <summary>
+        /// GetDumpFile will return empty list of strings if proc dump never started
+        /// </summary>
+        [TestMethod]
+        public void GetDumpFileWillReturnEmptyIfProcDumpDidntStart()
```
Comment (singhsarab) [inline test/Microsoft.TestPlatform.Extensions.BlameDataCollector.UnitTests/ProcessDumpUtilityTests.cs]: >GetDumpFileWillReturnEmptyIfProcDumpDidntStart [](start = 20, length = 46)

Exception?

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+                if (EqtTrace.IsVerboseEnabled)
+                {
+                    EqtTrace.Verbose("Using procdump at: {0}", procDumpExe);
+                }
+
+                return procDumpExe;
+            }
+            else
+            {
+                throw new Exception(Resources.Resources.ProcDumpEnvVarEmpty);
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs]: Add test for exception string

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+
+                if (this.environment.Architecture == PlatformArchitecture.X64)
+                {
+                    filename = "procdump64.exe";
+                }
+                else if (this.environment.Architecture == PlatformArchitecture.X86)
+                {
+                    filename = "procdump.exe";
+                }
+
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs]: Tests?

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+            this.processHelper.WaitForProcessExit(this.procDumpProcess);
+
+            var dumpFile = Path.Combine(this.testResultsDirectory, this.dumpFileName);
+            if (this.fileHelper.Exists(dumpFile))
+            {
+                return dumpFile;
+            }
+
+            if (EqtTrace.IsErrorEnabled)
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/ProcessDumpUtility.cs]: Reason: Exit code?

File: test/Microsoft.TestPlatform.Extensions.BlameDataCollector.UnitTests/ProcessDumpUtilityTests.cs [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+        {
+            Environment.SetEnvironmentVariable("PROCDUMP_PATH", null);
+
+            var processDumpUtility = new ProcessDumpUtility(this.mockProcessHelper.Object, this.mockFileHelper.Object, this.mockPlatformEnvironment.Object);
+
+            var ex = Assert.ThrowsException<TestPlatformException>(() => processDumpUtility.StartProcessDump(1234, "guid", "D:\\"));
+            Assert.AreEqual(ex.Message, Resources.Resources.ProcDumpEnvVarEmpty);
+        }
+
+        public void StartProcessDumpWillStartExeCorrespondingToPlatformArchitecture()
```
Comment (singhsarab) [inline test/Microsoft.TestPlatform.Extensions.BlameDataCollector.UnitTests/ProcessDumpUtilityTests.cs]: TestMethod ? :O

**Review by singhsarab — APPROVED** [review]

**Comment by singhsarab** [comment]
@dotnet-bot test this

**Comment by singhsarab** [comment]
@dotnet-bot help

**Comment by singhsarab** [comment]
@dotnet-bot test Windows_NT / Debug Build

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs [hunk 24]
```diff
@@ -133,15 +151,66 @@ private void SessionEnded_Handler(object sender, SessionEndEventArgs args)
...
+                    var dumpFile = this.processDumpUtility.GetDumpFile();
+                    if (!string.IsNullOrEmpty(dumpFile))
+                    {
+                        var fileTranferInformation = new FileTransferInformation(this.context.SessionDataCollectionContext, dumpFile, true);
+                        this.dataCollectionSink.SendFileAsync(fileTranferInformation);
+                    }
+                    else
+                    {
+                        if (EqtTrace.IsWarningEnabled)
+                        {
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs]: Not required.

**Pull request merged**